pkg/services/updatechecker/plugins.go

```diff
@@ -0,0 +1,171 @@
+package updatechecker
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-version"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type PluginsService struct {
+	availableUpdates map[string]string
+
+	enabled        bool
+	grafanaVersion string
+	pluginStore    plugins.Store
+	httpClient     httpClient
+	mutex          sync.RWMutex
+	log            log.Logger
+}
+
+func ProvidePluginsService(cfg *setting.Cfg, pluginStore plugins.Store) *PluginsService {
+	return &PluginsService{
+		enabled:          cfg.CheckForUpdates,
+		grafanaVersion:   cfg.BuildVersion,
+		httpClient:       &http.Client{Timeout: 10 * time.Second},
+		log:              log.New("plugins.update.checker"),
+		pluginStore:      pluginStore,
+		availableUpdates: make(map[string]string),
+	}
+}
+
+type httpClient interface {
+	Get(url string) (resp *http.Response, err error)
+}
+
+func (s *PluginsService) IsDisabled() bool {
+	return !s.enabled
+}
+
+func (s *PluginsService) Run(ctx context.Context) error {
+	s.checkForUpdates(ctx)
+
+	ticker := time.NewTicker(time.Minute * 10)
+	run := true
+
+	for run {
+		select {
+		case <-ticker.C:
+			s.checkForUpdates(ctx)
+		case <-ctx.Done():
+			run = false
+		}
+	}
+
+	return ctx.Err()
+}
+
+func (s *PluginsService) HasUpdate(ctx context.Context, pluginID string) (string, bool) {
+	s.mutex.RLock()
+	update, updateAvailable := s.availableUpdates[pluginID]
+	s.mutex.RUnlock()
+	if updateAvailable {
+		// check if plugin has already been updated since the last invocation of `checkForUpdates`
+		plugin, exists := s.pluginStore.Plugin(ctx, pluginID)
+		if !exists {
+			return "", false
+		}
+
+		// can ignore update as the update data is stale
+		if !canUpgrade(plugin.Info.Version, update) {
+			s.mutex.Lock()
+			delete(s.availableUpdates, pluginID)
+			s.mutex.Unlock()
+
+			return "", false
+		}
+
+		return update, true
```

Comment:
Wouldn't just return canUpgrade be enough here, not sure I understand the use case. But with my suggestion would create less lock contention and comparing versions over and over again doesn't feel like it will impact performance.

Reply:
Yeah I think that should be enough actually. Original thought was that if we know here that we can't upgrade, we should just remove the (now stale) update we found in the map